Socket connections from non-gradle applications crash gradle executors

### Current Behavior

A Gradle executor experiences this exception:
```
Unexpected exception thrown.
org.gradle.internal.remote.internal.MessageIOException: Could not read message from '/127.0.0.1:49184'.
        at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:94)
        at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:270)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.IllegalArgumentException: 
        at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:72)
        at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:52)
        at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:81)
        at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:270)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:750)
org.gradle.internal.remote.internal.ConnectException: Could not connect to server [4f40fbf1-8d30-4c4a-9ba1-aaa9b723f4ea port:39287, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1].
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:103)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: java.net.ConnectException: Connection refused
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:716)
	at sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:122)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54)
	... 5 more
```
And the build fails due to this error:
```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':storage:test'.
> Process 'Gradle Test Executor 102' finished with non-zero exit value 1
  This problem might be caused by incorrect test process configuration.
  For more on test execution, please refer to https://docs.gradle.org/8.5/userguide/java_testing.html#sec:test_execution in the Gradle documentation.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Get more help at https://help.gradle.org.
```

### Expected Behavior

The connection should be accepted, and then disconnected once the received data is shown to be invalid. The Gradle executor should continue running and accepting connections until a valid connection has been made. The Gradle executor could keep running indefinitely without a connection, or given a moderate (1min) timeout before the connection is abandoned completely.

### Context (optional)

We use Gradle to build the [Apache Kafka](https://github.com/apache/kafka) project, and received this [bug report](https://issues.apache.org/jira/browse/KAFKA-15343) from one of our contributors.

We have numerous tests which appear to leak clients in this way, and it appears that the particulars of the Gradle protocol and the Kafka protocol happen to cause the Gradle end of the connection to exit ungracefully.

We are also working to reduce the number of leaked clients and network connections in our project to mitigate this failure, but it is very likely that there are downstream users of Kafka clients which could experience the same style of defect in their tests, and have similar build failures. And more generally, any sort of client could happen to connect to the Gradle executor and trigger this failure with a crafted response. This could be happening without Kafka in use at all, theoretically.

This appears similar to the descriptions of #27512, #23602, and #12566 but I thought to open this issue separately with my reproduction case if it turns out to be distinct.

### Steps to Reproduce

1. A test opens a server socket with port 0, and gets assigned a random port N
2. The test starts a client which will continue to reconnect to port N indefinitely if the connection with the server is lost
3. The test stops the server, and port N is released to the OS and allowed to be reassigned
4. The client beings retrying connections to port N
5. A Gradle executor is eventually assigned port N by the OS
6. The client connects to port N, and reaches the Gradle executor unintentionally
7. The Gradle executor is killed due to an exception

Here is a reproducer: https://github.com/gharris1727/gradle-crash-reproducer . For a specific failed build, see https://github.com/gharris1727/gradle-crash-reproducer/actions/runs/7632481625/job/20792782276 .

The reproducer has 4 modules, structured such that the tests for `app-2` start after the tests for `app-1` are already running. `app-1` is continuously searching for ports which it can't bind() the address for, and connecting to them and then immediately closing the connection. This specific reproducer relies on MacOS for sequential port allocation in order to make it more reliable, but I think the same failure is possible on any platform.

Here's the logs from the reproducer, in case the action becomes inaccessible:
```
2024-01-23T22:13:08.059+0000 [DEBUG] [TestEventLogger] Gradle Test Run :app-2:test STARTED
...
2024-01-23T22:13:08.069+0000 [DEBUG] [org.gradle.internal.remote.internal.inet.TcpIncomingConnector] Listening on [77f73e7b-bbe9-4630-8033-0a25e66a99f3 port:58112, addresses:[localhost/127.0.0.1]].
...
2024-01-23T22:13:08.393+0000 [DEBUG] [TestEventLogger] TestWithLeakedClient > test() STANDARD_OUT
2024-01-23T22:13:08.393+0000 [DEBUG] [TestEventLogger]     [58112, 58113, 54322, 54323, 55555, 54324, 55556]
...
Error: -23T22:13:08.406+0000 [ERROR] [org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection] Unexpected exception thrown.
org.gradle.internal.remote.internal.MessageIOException: Could not write '/127.0.0.1:54325'.
	at org.gradle.internal.remote.internal.inet.SocketConnection.flush(SocketConnection.java:140)
	at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionDispatch.run(MessageHub.java:333)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.io.IOException: Broken pipe
	at java.base/sun.nio.ch.FileDispatcherImpl.write0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:47)
	at java.base/sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:113)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:58)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:50)
	at java.base/sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:462)
	at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.writeWithNonBlockingRetry(SocketConnection.java:279)
	at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.writeBufferToChannel(SocketConnection.java:267)
	at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.flush(SocketConnection.java:261)
	at org.gradle.internal.remote.internal.inet.SocketConnection.flush(SocketConnection.java:138)
	... 6 more
Error: -23T22:13:08.609+0000 [ERROR] [system.err] org.gradle.internal.remote.internal.ConnectException: Could not connect to server [77f73e7b-bbe9-4630-8033-0a25e66a99f3 port:58112, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1].
Error: -23T22:13:08.610+0000 [ERROR] [system.err] 	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67)
Error: -23T22:13:08.610+0000 [ERROR] [system.err] 	at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36)
Error: -23T22:13:08.610+0000 [ERROR] [system.err] 	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:103)
Error: -23T22:13:08.610+0000 [ERROR] [system.err] 	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
Error: -23T22:13:08.610+0000 [ERROR] [system.err] 	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
Error: -23T22:13:08.610+0000 [ERROR] [system.err] 	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Error: -23T22:13:08.611+0000 [ERROR] [system.err] Caused by: java.net.ConnectException: Connection refused
Error: -23T22:13:08.611+0000 [ERROR] [system.err] 	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
Error: -23T22:13:08.611+0000 [ERROR] [system.err] 	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:774)
Error: -23T22:13:08.611+0000 [ERROR] [system.err] 	at java.base/sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:120)
Error: -23T22:13:08.612+0000 [ERROR] [system.err] 	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81)
Error: -23T22:13:08.612+0000 [ERROR] [system.err] 	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54)
Error: -23T22:13:08.612+0000 [ERROR] [system.err] 	... 5 more
```

### Gradle version

8.5

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

--- a/platforms/core-runtime/launcher/src/integTest/groovy/org/gradle/launcher/daemon/DaemonInitialCommunicationFailureIntegrationSpec.groovy
+++ b/platforms/core-runtime/launcher/src/integTest/groovy/org/gradle/launcher/daemon/DaemonInitialCommunicationFailureIntegrationSpec.groovy
@@ -16,6 +16,7 @@
 
 package org.gradle.launcher.daemon
 
+import org.gradle.internal.remote.internal.inet.TcpOutgoingConnector
 import org.gradle.test.preconditions.UnitTestPreconditions
 import org.gradle.testdistribution.LocalOnly
 import org.gradle.integtests.fixtures.daemon.DaemonIntegrationSpec
@@ -31,6 +32,7 @@ import spock.lang.Issue
 import java.nio.ByteBuffer
 import java.nio.channels.ServerSocketChannel
 import java.nio.channels.SocketChannel
+import java.nio.charset.StandardCharsets
 
 import static org.gradle.test.fixtures.ConcurrentTestUtil.poll
 
@@ -164,6 +166,8 @@ class DaemonInitialCommunicationFailureIntegrationSpec extends DaemonIntegration
 
         when:
         def socket = new Socket(new InetAddressFactory().localBindingAddress, daemon.port)
+
+        socket.outputStream.write(TcpOutgoingConnector.CONNECTION_PREAMBLE)
         socket.outputStream.write("GET / HTTP/1.0\n\n".getBytes())
         socket.outputStream.flush()
 
@@ -198,17 +202,13 @@ class DaemonInitialCommunicationFailureIntegrationSpec extends DaemonIntegration
                 @Override
                 void run() {
                     while (true) {
-                        SocketChannel connection
                         try {
-                            connection = socket.accept()
+                            SocketChannel connection = socket.accept()
+                            connection.write(ByteBuffer.wrap("hello".getBytes(StandardCharsets.UTF_8)))
+                            //leave the connection open, so the client can read our message even if there is some delay
+                            //all connections are closed with the socket after the test
                         } catch (IOException e) {
                             return
-                        }
-                        try {
-                            connection.read(ByteBuffer.allocate(4096))
-                            connection.write(ByteBuffer.wrap("hello".bytes))
-                        } finally {
-                            connection.close()
                         }
                     }
                 }

--- a/platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonTcpServerConnector.java
+++ b/platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonTcpServerConnector.java
@@ -51,7 +51,8 @@ public class DaemonTcpServerConnector implements DaemonServerConnector {
         this.incomingConnector = new TcpIncomingConnector(
                 executorFactory,
                 inetAddressFactory,
-                new UUIDGenerator()
+                new UUIDGenerator(),
+                10
         );
     }
 

--- a/platforms/core-runtime/messaging/src/main/java/org/gradle/internal/remote/internal/inet/TcpIncomingConnector.java
+++ b/platforms/core-runtime/messaging/src/main/java/org/gradle/internal/remote/internal/inet/TcpIncomingConnector.java
@@ -29,25 +29,33 @@ import org.gradle.internal.remote.internal.IncomingConnector;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.IOException;
 import java.net.InetAddress;
 import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
 import java.nio.channels.ClosedChannelException;
 import java.nio.channels.ServerSocketChannel;
 import java.nio.channels.SocketChannel;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+import static org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.CONNECTION_PREAMBLE;
 
 public class TcpIncomingConnector implements IncomingConnector {
     private static final Logger LOGGER = LoggerFactory.getLogger(TcpIncomingConnector.class);
     private final ExecutorFactory executorFactory;
     private final InetAddressFactory addressFactory;
     private final IdGenerator<UUID> idGenerator;
+    private final int acceptTimeoutSeconds;
 
-    public TcpIncomingConnector(ExecutorFactory executorFactory, InetAddressFactory addressFactory, IdGenerator<UUID> idGenerator) {
+    public TcpIncomingConnector(ExecutorFactory executorFactory, InetAddressFactory addressFactory, IdGenerator<UUID> idGenerator, int acceptTimeoutSeconds) {
         this.executorFactory = executorFactory;
         this.addressFactory = addressFactory;
         this.idGenerator = idGenerator;
+        this.acceptTimeoutSeconds = acceptTimeoutSeconds;
     }
 
     @Override
@@ -109,13 +117,21 @@ public class TcpIncomingConnector implements IncomingConnector {
                         InetSocketAddress remoteSocketAddress = (InetSocketAddress) socket.socket().getRemoteSocketAddress();
                         InetAddress remoteInetAddress = remoteSocketAddress.getAddress();
                         if (!allowRemote && !addressFactory.isCommunicationAddress(remoteInetAddress)) {
-                            LOGGER.error("Cannot accept connection from remote address {}.", remoteInetAddress);
+                            LOGGER.error("Cannot accept connection from remote address {}.", remoteSocketAddress);
                             socket.close();
                             continue;
                         }
-                        LOGGER.debug("Accepted connection from {} to {}.", socket.socket().getRemoteSocketAddress(), socket.socket().getLocalSocketAddress());
                         try {
                             SocketBlockingUtil.configureNonblocking(socket);
+                            waitForConnectionPreamble(socket);
+                        } catch (IOException e) {
+                            LOGGER.error("Failed connection handshake with {}.", remoteSocketAddress, e);
+                            socket.close();
+                            continue;
+                        }
+
+                        LOGGER.debug("Accepted connection from {} to {}.", socket.socket().getRemoteSocketAddress(), socket.socket().getLocalSocketAddress());
+                        try {
                             action.execute(new SocketConnectCompletion(socket));
                         } catch (Throwable t) {
                             socket.close();
@@ -129,6 +145,23 @@ public class TcpIncomingConnector implements IncomingConnector {
                 }
             } finally {
                 CompositeStoppable.stoppable(serverSocket).stop();
+            }
+        }
+
+        private void waitForConnectionPreamble(SocketChannel socket) throws IOException, InterruptedException {
+            ByteBuffer buffer = ByteBuffer.allocate(CONNECTION_PREAMBLE.length);
+            long deadline = System.nanoTime() + TimeUnit.SECONDS.toNanos(acceptTimeoutSeconds);
+            while (buffer.hasRemaining() && System.nanoTime() < deadline) {
+                int read = socket.read(buffer);
+                if (read == -1) {
+                    break;
+                }
+                if (read == 0) {
+                    Thread.sleep(1);
+                }
+            }
+            if (!Arrays.equals(buffer.array(), CONNECTION_PREAMBLE)) {
+                throw new IOException("Did not receive connection preamble within " + acceptTimeoutSeconds + "s");
             }
         }
 

--- a/platforms/core-runtime/messaging/src/main/java/org/gradle/internal/remote/internal/inet/TcpOutgoingConnector.java
+++ b/platforms/core-runtime/messaging/src/main/java/org/gradle/internal/remote/internal/inet/TcpOutgoingConnector.java
@@ -31,10 +31,13 @@ import java.net.Socket;
 import java.net.SocketAddress;
 import java.net.SocketException;
 import java.net.SocketTimeoutException;
+import java.nio.ByteBuffer;
 import java.nio.channels.SocketChannel;
+import java.nio.charset.StandardCharsets;
 import java.util.List;
 
 public class TcpOutgoingConnector implements OutgoingConnector {
+    static final byte[] CONNECTION_PREAMBLE = "Gradle Magic".getBytes(StandardCharsets.UTF_8);
     private static final Logger LOGGER = LoggerFactory.getLogger(TcpOutgoingConnector.class);
     private static final int CONNECT_TIMEOUT = 10000;
 
@@ -84,9 +87,9 @@ public class TcpOutgoingConnector implements OutgoingConnector {
         SocketChannel socketChannel = SocketChannel.open();
         try {
             socketChannel.socket().connect(new InetSocketAddress(candidate, address.getPort()), CONNECT_TIMEOUT);
-
             if (!detectSelfConnect(socketChannel)) {
                 SocketBlockingUtil.configureNonblocking(socketChannel);
+                socketChannel.write(ByteBuffer.wrap(CONNECTION_PREAMBLE));
                 return socketChannel;
             }
             socketChannel.close();

--- a/platforms/core-runtime/messaging/src/main/java/org/gradle/internal/remote/services/MessagingServices.java
+++ b/platforms/core-runtime/messaging/src/main/java/org/gradle/internal/remote/services/MessagingServices.java
@@ -62,7 +62,8 @@ public class MessagingServices implements ServiceRegistrationProvider {
         return new TcpIncomingConnector(
                 executorFactory,
                 inetAddressFactory,
-                idGenerator
+                idGenerator,
+                10
         );
     }
 

--- a/platforms/core-runtime/messaging/src/test/groovy/org/gradle/internal/remote/internal/inet/TcpConnectorTest.groovy
+++ b/platforms/core-runtime/messaging/src/test/groovy/org/gradle/internal/remote/internal/inet/TcpConnectorTest.groovy
@@ -42,7 +42,7 @@ class TcpConnectorTest extends ConcurrentSpec {
     final def idGenerator = new UUIDGenerator()
     final def addressFactory = new InetAddressFactory()
     final def outgoingConnector = new TcpOutgoingConnector()
-    final def incomingConnector = new TcpIncomingConnector(executorFactory, addressFactory, idGenerator)
+    final def incomingConnector = new TcpIncomingConnector(executorFactory, addressFactory, idGenerator, 1)
     @Rule
     public ReleasingPortAllocator portAllocator = new ReleasingPortAllocator()
 
@@ -313,13 +313,39 @@ class TcpConnectorTest extends ConcurrentSpec {
         when:
         socketChannel.socket().bind(bindAnyPort)
         socketChannel.socket().connect(connectAddress)
+        socketChannel.socket().outputStream.write(TcpOutgoingConnector.CONNECTION_PREAMBLE)
+        socketChannel.socket().outputStream.flush()
 
         then:
         !outgoingConnector.detectSelfConnect(socketChannel)
 
         cleanup:
         acceptor?.stop()
-        socketChannel.close()
+        socketChannel?.close()
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/27801")
+    def "refuses connections that don't provide a handshake"() {
+        given:
+        def connected = false
+        def action = { connected = true  }
+        def socketChannel = SocketChannel.open()
+        def acceptor = incomingConnector.accept(action, false)
+        def communicationAddress = addressFactory.getLocalBindingAddress()
+        def bindAnyPort = new InetSocketAddress(communicationAddress, 0)
+        def connectAddress = new InetSocketAddress(communicationAddress, acceptor.address.port)
+
+        when:
+        socketChannel.socket().bind(bindAnyPort)
+        socketChannel.socket().connect(connectAddress)
+        acceptor.stop()
+
+        then:
+        !connected
+
+        cleanup:
+        acceptor?.stop()
+        socketChannel?.close()
     }
 }
 

